Feature/kebastian

sebastianasen opened the pull request:
Fixed a few things in the Controller and Cluster.
Added a new State 'Charging finished' that is set if the car is full (The Keba hasn't an own state for finishing the charging, so a few conditions must fullfilled).

sfeilmeier commented on io.openems.edge.controller.evcs/src/io/openems/edge/controller/evcs/EvcsController.java:
```diff
@@ -302,4 +321,21 @@ public ModbusSlaveTable getModbusSlaveTable(AccessMode accessMode) {
+				c.update(properties);
+			}
+		} catch (IOException | SecurityException e) {
+			System.out.println("ERROR: " + e.getMessage());
```
Verwende hier die `this.logError()` Methode

sfeilmeier commented on io.openems.edge.evcs.api/src/io/openems/edge/evcs/api/Status.java:
```diff
@@ -16,7 +16,8 @@
 	CHARGING_REJECTED(5, "Charging rejected"),
-	ENERGY_LIMIT_REACHED(6, "The charging limit reached");
+	ENERGY_LIMIT_REACHED(6, "The charging limit reached"),
+    CHARGING_FINISHED(7, "Charging has finished");
```
Eclipse Autoformat

sfeilmeier commented on io.openems.edge.evcs.keba.kecontact/src/io/openems/edge/evcs/keba/kecontact/core/KebaKeContactCoreImpl.java:
```diff
@@ -49,17 +48,17 @@ public void onReceive(BiConsumer<InetAddress, String> callback) {
 	private class ReceiveWorker extends AbstractImmediateWorker {
 
-		private final DatagramSocket socket;
+		private DatagramSocket socket;
```
Warum nicht mehr final?

sebastianasen replied:
Das hat durch die änderungen an der deactivate Methode Fehler verurscht.

sfeilmeier commented on io.openems.edge.simulator/src/io/openems/edge/simulator/evcs/SimulatedEvcs.java:
```diff
@@ -21,18 +23,23 @@
-public class SimulatedEvcs extends AbstractOpenemsComponent implements ManagedEvcs, OpenemsComponent, EventHandler {
+public class SimulatedEvcs extends AbstractOpenemsComponent implements ManagedEvcs, Evcs, OpenemsComponent, EventHandler {
 
+	public LocalDateTime startTime;
```
Warum public?

sfeilmeier commented on io.openems.edge.simulator/src/io/openems/edge/simulator/evcs/SimulatedEvcs.java:
```diff
@@ -84,10 +102,13 @@ public void handleEvent(Event event) {
 		Optional<Integer> chargePowerLimitOpt = this.setChargePowerLimit().getNextWriteValueAndReset();
-
+		
+		System.out.println("chargePowerLimitOpt"+chargePowerLimitOpt);
```
Alle system.out.println sollten hier weg

sfeilmeier commented on io.openems.edge.simulator/src/io/openems/edge/simulator/evcs/SimulatedEvcs.java:
```diff
@@ -84,10 +102,13 @@ public void handleEvent(Event event) {
 		}
 	}
-
+	private LocalDateTime letzterUpdate = LocalDateTime.now();
```
"lastUpdate"

sfeilmeier commented on io.openems.edge.simulator/src/io/openems/edge/simulator/evcs/SimulatedEvcs.java:
```diff
@@ -100,8 +121,18 @@ private void updateChannels() {
+		long timeDiff = ChronoUnit.MILLIS.between(letzterUpdate, LocalDateTime.now());
+		double energieTransfered = (timeDiff / 1000.0/60/60) * this.getChargePower().getNextValue().orElse(0);
+		System.out.println(this.getChargePower().getNextValue().orElse(0));
+		exactEnergySession = exactEnergySession + energieTransfered;
```
Zugriff mit `this.` damit klar wird, dass das eine Klassenvariable ist

sfeilmeier commented on ui/src/app/edge/live/consumption/modal/modal.component.html:
```diff
@@ -23,7 +23,7 @@
           </table>
           <!-- for spacing between Phases and Total Consumption -->
-          <div *ngIf="sum.activePowerL1 != null && sum.activePowerL2 != null && sum.activePowerL3 != null"
+          <div ngIf="sum.activePowerL1 != null && sum.activePowerL2 != null && sum.activePowerL3 != null"
```
Ist der Stern nicht notwendig?

sfeilmeier commented on ui/src/app/edge/live/evcs/modal/popover/evcs-popover.page.html:
```diff
@@ -0,0 +1,86 @@
+        <table class="full_width">
+            <tr>
+                <td style="text-align: center">
+                    Kapazitätsorientierte Beladung
```
Base-Language ist Englisch

sebastianasen commented:
Fixed a few things in the Controller and Cluster.
Added a new State 'Charging finished' that is set if the car is full.
Changed a few things in the UI.
